### DISCUSSION TITLE
Vanilla integration

### DISCUSSION THREAD

**Discussion by Maygician** [yume-chan/ya-webadb#371] [Q&A]
Need separate feature bundles, that are easy to integrate in a vanilla project

**Comment by Maygician** [comment]
Or I'd appreciate a few hints on how to make em into callable functions.

**Comment by yume-chan** [comment]
1. Install the core package [`@yume-chan/adb`](https://www.npmjs.com/package/@yume-chan/adb)
2. Install a backend for connection and a credential store for creating/retrieving ADB keys. 
    * For browser environment, [`@yume-chan/adb-backend-webusb`](https://www.npmjs.com/package/@yume-chan/adb-backend-webusb) provides WebUSB connection and [`@yume-chan/adb-credential-web`](https://www.npmjs.com/package/@yume-chan/adb-credential-web) provides key creation using WebCrypto and storage using LocalStorage. 
    * For Node.js, there is no official backend and credential store yet.
3. Each backend initialize differently, with `@yume-chan/adb-backend-webusb`:
    * Use `AdbWebUsbBackend.requestDevice()` to request access to a new device
    * Use `AdbWebUsbBackend.getDevices()` to get list of connected and authorized devices
4. Create `Adb` instance by calling `const adb = new Adb(backend)`
5. Connect to device by calling `adb.connect(credentialStore)`
6. Invoke ADB commands by calling methods on `adb`

See https://github.com/yume-chan/ya-webadb/blob/4399fedff84229a1daf24d6aa62bd3168c19eabd/apps/demo/components/connect.tsx for example.

**Reply by Maygician** [reply]
Howd you bundle it? tsc throws bunch of syntax errors, and it transpiles ?. and ?? operators to weird code breaking stuff like nested emty ternary operators. I dont use npm and need ready to cdn bundles (actually adding in runtime)

**Reply by yume-chan** [reply]
@Maygician then check [CONTRIBUTE.md](https://github.com/yume-chan/ya-webadb/blob/4399fedff84229a1daf24d6aa62bd3168c19eabd/CONTRIBUTE.md) or [CI workflow](https://github.com/yume-chan/ya-webadb/blob/4399fedff84229a1daf24d6aa62bd3168c19eabd/.github/workflows/nodejs.yml#L31-L33), I guess you didn't initialize the monorepo.

I think you need to fork the repo, add a new package, reference all things you need, and configure the bundler. It doesn't require any special bundler options, a normal Webpack will work.

**Reply by adnanaga** [reply]
Is there a way to only pull the scrcpy component into it's own module. The ability to use it without installing anything on the machine itself and only through the browser is amazing

**Reply by yume-chan** [reply]
@adnanaga Yes, the Scrcpy client implementation is also published to NPM: https://www.npmjs.com/package/@yume-chan/scrcpy

**Reply by JulesLecuir** [reply]
Hi @yume-chan, I took a look at the file you gave us [here](https://github.com/yume-chan/ya-webadb/blob/4399fedff84229a1daf24d6aa62bd3168c19eabd/apps/demo/components/connect.tsx) but I can't get to make it work in my app.

I run a React JS web app, and so far, all I did is copy-paste your imports into an existing file of my app, just to check if that could work:
```js
import {Adb, AdbBackend} from "@yume-chan/adb";
import AdbWebUsbBackend, {AdbWebUsbBackendWatcher} from "@yume-chan/adb-backend-webusb";
import AdbWebCredentialStore from "@yume-chan/adb-credential-web";
```

But I get that error during the build:

```
./node_modules/@yume-chan/adb/esm/adb.js 102:12
Module parse failed: Unexpected token (102:12)
File was processed with these loaders:
 * ./node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|   }
| 
>   dispatcher;
| 
|   get disconnected() {
```

Maybe you already faced this kind of problem? Is any import missing? I found pretty much nothing on the internet about it. I use version 0.0.16 of your tools. 

Thank you so much !

**Reply by yume-chan** [reply]
I'm not a Webpack expert. Looks like you are using a pretty old version of Webpack which doesn't understand class fields. And you are transforming JS files in node_modules with babel, which is usually unnecessary (for latest Webpack version) and causes slower compilation.

If you really want to use this setup, configure your babel to downlevel transform to a even older ES version to get rid of the class fields.

**Reply by JulesLecuir** [reply]
Thanks, I will check that out!

**Reply by dave-bvr** [reply]
@yume-chan Thanks for the library! It's working great for me via the demo link as well as on my localhost, building with rush. When I try to integrate it into an existing project (React, built with Vite) I'm experiencing some issues. I'm able to pair and connect to different Android devices using your `adb-backend-webusb` and `adb-credential-web` modules, but just trying to `import { Adb } from '@yume-chan/adb'` and then instantiate it, throws this error at runtime:
```
dispatcher.ts:65 Uncaught TypeError: Cannot read properties of undefined (reading 'readable')
    at new AdbPacketDispatcher (dispatcher.ts:65:20)
    at new Adb (adb.ts:176:27)
    at ListDevices.jsx:19:13
AdbPacketDispatcher @ dispatcher.ts:65
Adb @ adb.ts:176
(anonymous) @ ListDevices.jsx:19
```

I can provide more detail about my local setup if it would be helpful. Thanks again!

**Reply by yume-chan** [reply]
@dave-bvr From your error stack you passed `undefined` into `new Adb`. Also you should only call `new Adb` if your connection is already authenticated (or doesn't need authentication, for example when using with a debug build of ADB daemon), call `Adb.authenticate` instead. Without your exact code I can't predict where it goes wrong, please open a new issue with related code if you still can't get it to work.

**Reply by dave-bvr** [reply]
Thanks @yume-chan! I actually think I was going wrong by following the example here: https://github.com/yume-chan/ya-webadb/blob/4399fedff84229a1daf24d6aa62bd3168c19eabd/apps/demo/components/connect.tsx#L133
but using it with the latest version of the npm module (0.0.17) in which adb's constructor expects 4 arguments instead of two. I rolled back to 0.0.10 and it seems to be working now.

**Comment by adnanaga** [comment]
Do you have an example on how to implement only scrcpy with clickable control ?

I’m only trying to use Webadb and scrcpy
